src/Microsoft.AspNet.Http.Core/Collections/RequestCookiesCollection.cs

```diff
@@ -73,12 +72,20 @@ public IList<string> GetValues(string key)
             return _dictionary.TryGetValue(key, out value) ? new[] { value } : null;
         }
 
-        private static readonly char[] SemicolonAndComma = { ';', ',' };
-
-        public void Reparse(string cookiesHeader)
+        public void Reparse(IList<string> values)
         {
             _dictionary.Clear();
-            ParsingHelpers.ParseDelimited(cookiesHeader, SemicolonAndComma, AddCookieCallback, _dictionary);
+
+            IList<CookieHeaderValue> cookies;
+            if (CookieHeaderValue.TryParseList(values, out cookies))
+            {
+                foreach (var cookie in cookies)
+                {
+                    var name = Uri.UnescapeDataString(cookie.Name.Replace('+', ' '));
```

Comment:
How many places do we need to do this? Can we use @GrabYourPitchforks  de/encoder?